codalab/worker/file_util.py

```diff
@@ -171,11 +184,192 @@ def do_unzip(filename):
+        tf, _ = open_indexed_tar_gz_file(linked_bundle_path.bundle_path)
+
+        isdir = lambda finfo: finfo.type == tarfile.DIRTYPE
+
+        fpath = "/" + linked_bundle_path.archive_subpath
+        finfo = tf.getFileInfo(fpath)
+        if finfo is None:
+            raise FileNotFoundError(fpath)
+        if isdir(finfo):
+            from codalab.worker.download_util import compute_target_info_beam_descendants_flat
+
+            # If streaming a folder within an Azure bundle, we need to download its contents,
+            # re-archive the folder, and return the .tar.gz version of that folder.
+            # with tempfile.TemporaryDirectory() as tmp_dirname:
+            tmp_dir = tempfile.TemporaryDirectory()
+            extracted_path = os.path.join(tmp_dir.name, linked_bundle_path.archive_subpath)
+            os.mkdir(extracted_path)
+            for member in compute_target_info_beam_descendants_flat(file_path):
+                # Make sure that there is no trickery going on (see note in
+                # TarFile.extractall() documentation.
+                member_path = os.path.realpath(os.path.join(extracted_path, member['name']))
+                if not member_path.startswith(os.path.realpath(extracted_path)):
+                    raise tarfile.TarError('Archive member extracts outside the directory.')
+                # Extract other members of the directory.
+                # TODO (Ashwin): Make sure this works with symlinks, too.
+                if member['type'] == 'directory':
+                    os.makedirs(member_path, exist_ok=True)
+                else:
+                    with open(member_path, "wb+") as f:
+                        os.makedirs(os.path.dirname(member_path), exist_ok=True)
+                        f.write(
+                            tf.read(
+                                path="/"
+                                + linked_bundle_path.archive_subpath
+                                + "/"
+                                + member['name'],
+                                size=member['size'],
+                                offset=0,
+                            )
+                        )
+            # We use ClosingStreamWrapper as a workaround to ensure that tmp_dir isn't automatically closed.
+            # If we just return `tar_gzip_directory(extracted_path)`, for some reason, Python closes
+            # tmp_dir, so we end up just reading from an empty file.
+            return ClosingStreamWrapper(
+                tar_gzip_directory(extracted_path), lambda t: t.cleanup(), (tmp_dir,),
+            )
+        else:
+            # TODO: Implement a tf.open() function so that we don't have to read the entire file.
+            return BytesIO(tf.read(path="", fileInfo=finfo, size=finfo.size, offset=0))
+    return FileSystems.open(file_path, mode, compression_type=CompressionTypes.UNCOMPRESSED)
+
+
+class GzipStream:
```

Comment:
I thought the Python version of this is not performant and that's why we call out to `gzip`?

Reply:
Yes -- but we can't easily use the command line `gzip` to read from a Python file-like object (earlier we would just be passing a file path to `gzip`). See comment on line 383:

```
        # We run gzip in the same process if the file is on Azure, so that
        # we can use Apache Beam methods to read the file. This may cause
        # performance issues once we switch all files to use Azure, but
        # they should be mitigated when we allow users to connect to Azure
        # servers directly to download files.
```

Potentially it might be possible to read from the Python file-like object and pipe that into stdin to the `gzip` process, instead -- do you think that might be worth trying out?

Reply:
For simplicity, I'm just going to keep using GzipStream in this PR. We can think about doing what I mentioned (pipe into stdin for `gzip`) in another PR as that would add additional complexity to everything.